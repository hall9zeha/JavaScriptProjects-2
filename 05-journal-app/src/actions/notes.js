import { db } from "../firebase/firebaseConfig";
import { types } from "../types/Types";

export const startNewNote = () =>{
    // getState es el segundo parámetro proveido por redux-thunk que nos permite despachar las acciones
    // y funciona muy parecido a useSelector al acceder a los objetos de nuestros estados en redux, para 
    // extraer en este caso los datos de usuario en sesión.
    return async (dispatch,getState)=>{
        const {uid} = getState().auth;

        const newNote = {
            title:'',
            body:'',
            date: new Date().getTime()
        }
        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote)
        // Un vez creada la nota en firebase solo necesitaremos su id y enviaremos nuevamente el objeto newNote
        // que está vacío hacia nuestra vista, para actualizar los cambios que hagamos y guardarlos
        dispatch(activeNote(docRef.id,newNote))
    }   
        
}

export const activeNote = (id, note) =>({
    type:types.notesActive,
    payload:{
        id,
        ...note
    }
})

export const setNotes = (notes) =>({
    type:types.notesLoad,
    payload:notes
})