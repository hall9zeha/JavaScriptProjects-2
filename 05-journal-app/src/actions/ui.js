import { types } from "../types/Types";

export const setError =  (error) =>({
    type:types.uiSetError,
    payload:error
});

export const removeError = () =>({
    type:types.uiRemoveError
})

export const uiStartLoading =()=>({
    type:types.uiStartLoading,
    payload:true
})
export const uiFinishLoading = ()=>({
    type:types.uifinishLoading,
    payload:false
})