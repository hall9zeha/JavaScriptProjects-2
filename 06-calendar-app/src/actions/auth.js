import Swal from "sweetalert2";
import { fetchWithToken, fetchWithoutToken } from "../helpers/fetch"
import { types } from "../types/types";

export const startLogin = (email, password) =>{
    //Gracias a la librería thunk podemos disponer de un dispatch
    return async(dispatch) =>{
        const resp = await fetchWithoutToken('auth',{email,password},'POST');
        const body = await resp.json();
        
        //Si recibimos un arespuesta correcta de nuestra api 
        /*
            {
                ok:true
                ...
            }
        guardamos el token y la fecha de creación en local storage
        */
       
        if(body.ok){
            localStorage.setItem('token',body.token);
            localStorage.setItem('token-init-date',new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))
        }else{
            Swal.fire('Error',body.msg, 'error');
        }
    }
}

export const startRegister = (name,email,password) =>{
    return async(dispatch)=>{
        const resp = await fetchWithoutToken('auth/new',{name,email,password},'POST');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
        }

    }
}
export const startChecking =()=> {
    return async(dispatch)=>{

        //No requiere ningún argumento porque por defecto será una petición GET
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
            dispatch(checkingFinish());
        }
    }
}

const checkingFinish = () =>({type:types.authCheckingFinish});
//Ya que usaremos esta acción solo en este lugar no la exportaremos
const login = (user) =>({
    type:types.authLogin,
    payload:user
})

