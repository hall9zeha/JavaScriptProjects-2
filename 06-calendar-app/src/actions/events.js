import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch"
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types"

//Si se guardó nuestro evento en la base de datos mongo en la nube entonces disparar eventAddNew()
export const eventStartAddNew = (event)=>{
    return async(dispatch, getState)=>{
        //Obtenemos el nombre, uid del usuario que ya está en nuestro store de redux
        const {uid, name} = getState().auth;
        try {
            
            const resp = await fetchWithToken('events',event,'POST');
            const body = await resp.json();
            console.log(body)
            if(body.ok){
                event.id=body.savedEvent.id;
                event.user = {
                    _id:uid,
                    name:name
                }
                
                dispatch(eventAddNew(event))
            }
        
        } catch (error) {
            console.log(error)
        }
    }
}

const eventAddNew =(event)=>({
    type:types.eventAddNew,
    payload:event
})

export const eventSetActive =(event)=>({
    type:types.eventSetActive,
    payload:event
})
export const eventClearActiveEvent = () =>({type:types.eventClearActiveEvent})

export const eventStartUpdate =(event) =>{
    return async(dispatch)=>{
        try {
            const resp = await fetchWithToken(`events/${event.id}`,event,'PUT');
            const body = await resp.json();

            if(body.ok){
                dispatch(eventUpdated(event));
            }else{
                Swal.fire('Error',body.msg,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventUpdated =(event)=>({
    type:types.eventUpdated,
    payload:event
})

export const eventStartDelete =()=>{
    return async(dispatch,getState)=>{
        const {id} = getState().calendar.activeEvent;
        
        try {
            const resp = await fetchWithToken(`events/${id}`,{},'DELETE');
            const body = await resp.json();

            if(body.ok){
                dispatch(eventDeleted());
              
            }else{
                Swal.fire('Error',body.msg,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventDeleted = ()=>({
    type:types.eventDeleted
})

export const eventStartLoading = ()=>{

    return async(dispatch,getState)=>{
        const {uid} = getState().auth;
        try {
            const resp = await fetchWithToken('events');
            const body = await resp.json();
            const events = prepareEvents(uid,body.events);
            dispatch(eventLoaded(events));

        } catch (error) {
            
        }
    }
}

const eventLoaded=(events)=>({
    type:types.eventLoaded,
    payload:events
})

export const eventLogout=()=>({
    type:types.eventLogout
})