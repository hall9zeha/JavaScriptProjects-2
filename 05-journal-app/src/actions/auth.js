// Las acciones no son más que simples funciones que representan nuestros casos de uso

import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/Types";
import { uiFinishLoading, uiStartLoading } from "./ui";



export const startLoginWithEmailPassword = (email, password) =>{
    return (dispatch)=>{
        dispatch(uiStartLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName));
                dispatch(uiFinishLoading())
            })
            .catch(e=>{
                dispatch(uiFinishLoading())
                console.log(e)
            })
        
    }
}

export const startLoginWithGoogle = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName))
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async({user}) =>{
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid,user.displayName));
            })
            .catch(e=>{
                console.log(e)
            })
    }
}


// La siguiente función retornará un objeto
export const login = (uid, displayName)=>({
    type: types.login,
    payload:{
        uid,
        displayName
    }
})

// Logout

export const startLogout = ()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout=()=>({
    type:types.logout
})