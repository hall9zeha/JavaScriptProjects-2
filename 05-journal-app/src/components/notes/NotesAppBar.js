import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const {active:currentNote} = useSelector(state=>state.notes)

  const handleSave=()=>{
    dispatch(startSaveNote(currentNote))
  }

  const handlePictureUpload = ()=>{
    document.querySelector('#fileSelector').click()
  }
  const handleFileChange = (e)=>{
    
    const file = e.target.files[0]
    if(file){
      dispatch(startUploading(file))
    }
  }
  
  
  return (
    <div className='notes__appbar'>
        <span>13 julio 2024</span>
        <input
          id='fileSelector'
          type='file'
          name='file'
          style={{display:'none'}}
          onChange={handleFileChange}
        />
        <div>
            <button 
              onClick={handlePictureUpload}
              className='btn'>Picture</button>
            <button 
              onClick={handleSave}
              className='btn'>Save</button>
        </div>
    </div>
  )
}
