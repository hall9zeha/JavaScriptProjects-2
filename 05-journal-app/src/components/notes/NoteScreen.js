import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { activeNote } from '../../actions/notes'

export const NoteScreen = () => {

    const dispatch = useDispatch();
    //Renombramos el objeto desestructurado 'active' a 'note' active:note
    const {active:note} = useSelector(state =>state.notes)
    const [formValues, handleInputChange, reset] = useForm(note);
    const {body, title} = formValues

    /*
    Para que los campos del formulario de nuestra nota cambien al seleccionar otra
    debemos resetear el formulario con los nuevos valores, para ello usaremos lo siguiente
    
    useRef almacena una variable mutable y no va aredibujar el componente si esta cambia
    */
    const activeNoteId =  useRef(note.id);
    useEffect(()=>{
        //Si el id de la nota seleccionada es diference a nuestro id inicial guardado
        //entonces reetear el formulario y establecer activeNoteId con el nuevo valor
        if(note.id !== activeNoteId.current){
            reset(note);
            activeNoteId.current = note.id
        } 
    },[note,reset])

    //Otro useEffect para actualizar el estado de la nota activa cuando escribimos algo dentro de sus campos
    useEffect(()=>{
        dispatch(activeNote(formValues.id,{...formValues}))
    },[formValues,dispatch])

  return (
    <div className='notes__main-content'>
        <NotesAppBar/>
        <div className='notes__content'>
            <input 
                type='text'
                placeholder='Some awesome'
                className='notes__title-input'
                name='title'
                value={title}
                onChange={handleInputChange}
                />
            <textarea
                placeholder='What happened today'
                className='notes__textarea'
                name='body'
                value={body}
                onChange={handleInputChange} 
            
            ></textarea>
            {
                note.url &&
                    <div className='notes__image'>
                        <img
                        src={note.url}
                        alt='nature road'
                        />
                    </div>
            }
           
        </div>
    </div>
  )
}
