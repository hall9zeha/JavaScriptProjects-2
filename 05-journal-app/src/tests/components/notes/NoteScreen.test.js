/**
 * @jest-environment jsdom
 */
import {Provider} from 'react-redux';
import renderer ,{act}from 'react-test-renderer';


import configureStore from 'redux-mock-store'

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';


const thunk = require('redux-thunk').thunk;

jest.mock('../../../actions/notes',()=>({
    activeNote:jest.fn()
}))


//***************************/
const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
    auth:{
        id:'123',
        name:'Martha'
    },
    ui:{
        loading:false,
        msgError:null
    },
    notes:{
        active:{
            id:'Abs',
            title:'Hola',
            body:'Esta es una prueba'
        },
        notes:[]
    }
}
let store = mockStore(initState)
store.dispatch=jest.fn()

describe('Tests in <NoteScreen/> component', () => { 
    const wrapper = renderer.create(
        <Provider store={store}>
            <NoteScreen/>
        </Provider>
    )
    test('should show successfully', () => { 
        expect(wrapper).toMatchSnapshot()
     })
    test('should call activeNote actions when change a textInput value',()=>{
        const txtTitle = wrapper.root.findByProps({name:'title'})
        const value='Hello world'
        act(()=>{
            txtTitle.props.onChange({target:{
                name:'title',
                value:value
            }})
        })
       //Ya que se ejecuta dos veces: activeNote al cargar la nota y la segunda al realizar las modificaciones
       //Solo nos centraremos en la Ãºltima llamada
        expect(activeNote).toHaveBeenLastCalledWith('Abs',{
            body:'Esta es una prueba',
            title:'Hello world',
            id:'Abs'
        })
    })
 })