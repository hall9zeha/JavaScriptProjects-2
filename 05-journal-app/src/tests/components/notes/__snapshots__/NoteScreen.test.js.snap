// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <NoteScreen/> component should show successfully 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Thursday 14 2024
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome"
      type="text"
      value="Hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Esta es una prueba"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
